src/librustc_mir/transform/promote_consts.rs

```diff
@@ -200,14 +202,18 @@ impl<'tcx> Visitor<'tcx> for Collector<'_, 'tcx> {
     fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
         self.super_rvalue(rvalue, location);
 
-        match *rvalue {
+        match rvalue {
             Rvalue::Ref(..) => {
                 self.candidates.push(Candidate::Ref(location));
             }
-            Rvalue::Repeat(..) if self.tcx.features().const_in_array_repeat_expressions => {
-                // FIXME(#49147) only promote the element when it isn't `Copy`
-                // (so that code that can copy it at runtime is unaffected).
-                self.candidates.push(Candidate::Repeat(location));
+            Rvalue::Repeat(elem, _) if self.tcx.features().const_in_array_repeat_expressions => {
+                if !elem.ty(&self.body.local_decls, self.tcx).is_copy_modulo_regions(
```

Comment:
> This PR doesn't change the order in which Copy-check and promotion-check are done, does it? Where is that order determined?

It only performs the promotion check if the type is not `Copy`

Reply:
Ah I see, thanks.

I think this confirms that somebody else ought to review this, if I cannot even tell as much. ;)

Reply:
Almost all of this PR is plumbing to be able to call `is_copy_modulo_regions`, which is the main way you can check for `Copy` outside of the MIR borrowck.